daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/VersionRange.scala

```diff
@@ -3,7 +3,36 @@
 
 package com.daml.lf
 
+import com.daml.lf.transaction.VersionTimeline
+
 final case class VersionRange[V](
     min: V,
     max: V,
-)
+) {
+  import VersionRange._
+  import Ordering.Implicits._
+
+  def intersect(that: VersionRange[V])(
+      implicit ev: VersionTimeline.SubVersion[V]): VersionRange[V] = {
+    VersionRange(
+      min = this.min max that.min,
+      max = this.max min that.max
+    )
+  }
+
+  def nonEmpty(implicit ev: VersionTimeline.SubVersion[V]): Boolean =
+    min <= max
+
+  def contains(v: V)(implicit ev: VersionTimeline.SubVersion[V]): Boolean =
+    min <= v && v <= max
+
+}
+
+object VersionRange {
+
+  import VersionTimeline.Implicits._
+
+  private implicit def ordering[V](implicit ev: VersionTimeline.SubVersion[V]): Ordering[V] =
```

Comment:
This is why we define `precedes` syntax instead of an `Order` instance in the first place.

```suggestion
  // do not publish this instance; it is unlawful as it is not antisymmetric
  private implicit def ordering[V](implicit ev: VersionTimeline.SubVersion[V]): Ordering[V] =
```

Reply:
It was not clear to me that proceed is not antisymmetric. Thanks to point that to me. 
I understand now why you did not define an `Ordering`, and I agree that is better. 
Let's me drop the Ordering here.